### PR TITLE
#4636: Add support for empty mesh in AnalyzeMesh pass and add explicit issues for unsupported ops in multi device path.

### PR DESCRIPTION
**Pull request by tapspatel**
https://github.com/tenstorrent/tt-mlir/issues/4636

This translates any empty mesh into 1x1 mesh.

### CODE REVIEW

File: test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir [hunk 1]
```diff
@@ -1,13 +1,29 @@
 // REQUIRES: stablehlo
-// RUN: ttmlir-opt --stablehlo-pipeline -o %t.mlir %s
+// RUN: ttmlir-opt -split-input-file --stablehlo-pipeline -o %t.mlir %s
 // RUN: FileCheck %s --input-file=%t.mlir
 
-module @jit_reshape attributes {mhlo.num_partitions = 8 : i32, mhlo.num_replicas = 1 : i32} {
-  sdy.mesh @mesh = <["x"=1, "batch"=8]>
+// CHECK-LABEL: module {
```
Comment (mmanzoorTT) [inline test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir]: Input file is already split with `-split-input-file`, so I don't see a need for `CHECK-LABEL: module {`

File: test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir [hunk 2]
```diff
@@ -1,13 +1,29 @@
...
+module {
+  sdy.mesh @mesh = <["batch"=8]>
```
Comment (mmanzoorTT) [inline test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir:7]: What will happen for the old mesh `sdy.mesh @mesh = <["x"=1, "batch"=8]>`?

File: test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir [hunk 3]
```diff
@@ -1,13 +1,29 @@
...
   func.func public @main(%arg0: tensor<1024x2x32x32xf32> {sdy.sharding = #sdy.sharding<@mesh, [{"batch"}, {}, {}, {}]>}) -> (tensor<2048x1024xf32> {jax.result_info = ""}) {
     %0 = stablehlo.reshape %arg0 : (tensor<1024x2x32x32xf32>) -> tensor<2048x1024xf32>
     return %0 : tensor<2048x1024xf32>
   }
 }
 
-// CHECK: sdy.mesh @mesh = <["x"=1, "batch"=8]>
+// CHECK: sdy.mesh @mesh = <["batch_updated"=1, "batch"=8]>
```
Comment (mmanzoorTT) [inline test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir]: Suggestion: move it with `sdy.mesh @mesh = <["batch"=8]>`. It will make easy to follow the test.

File: test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir [hunk 4]
```diff
@@ -1,13 +1,29 @@
...
+
+// -----
+// CHECK-LABEL: module {
```
Comment (mmanzoorTT) [inline test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir]: Same suggestion for the `CHECK` statement as above.

File: lib/Dialect/StableHLO/Transforms/AnalyzeMesh.cpp [hunk 5]
```diff
@@ -259,8 +259,15 @@ class AnalyzeMeshPass : public impl::AnalyzeMeshPassBase<AnalyzeMeshPass> {
           shardy_utils::getMeshShapeFromMeshAttr(
               parsedMeshOps[0].getMeshAttr());
       llvm::SmallVector<int64_t> newMeshShape = originalMeshShape;
-      if (newMeshShape.size() == 1) {
+
+      if (newMeshShape.size() == 0) {
+        newMeshShape = {1, 1};
+      } else if (newMeshShape.size() == 1) {
         newMeshShape = {1, newMeshShape[0]};
+      } else if (newMeshShape.size() > 2) {
```
Comment (mmanzoorTT) [inline lib/Dialect/StableHLO/Transforms/AnalyzeMesh.cpp]: `newMeshSahpe.size() == 2` case is not covered.

Comment (tapspatel) [reply lib/Dialect/StableHLO/Transforms/AnalyzeMesh.cpp]: in that case, we don't need to change the mesh shape. These cases are only if we need to modify the mesh shape. I will write a comment about it.

File: test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir (hunk 2 above)
Comment (tapspatel) [reply test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir:7]: that was a mistake on my end when I first added the test, technically this test should be checking whether we bump up a 1D mesh to 2D (before it was already 2D).

File: test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir (hunk 3 above)
Comment (tapspatel) [reply test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir]: fixed

File: test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir (hunk 1 above)
Comment (tapspatel) [reply test/ttmlir/Dialect/StableHLO/analyze_mesh/shardy.mlir]: fixed
